File: src/morie/fn/aitclri.py
# morie.fn -- function file (rootcoder007/morie)
"""Inverse centred log-ratio transform."""

import math

from . import _tail1core as C

from ._richresult import RichResult

__all__ = ['clrinv', 'aitchison_clr_inverse']


def clrinv(z, total=1.0):
    """Inverse centred log-ratio transform.

    Formula: clr^-1(z) = C( exp(z_1), ..., exp(z_D) )

    Parameters
    ----------
    z : array-like
        Centred log-ratio coordinates, length D.
    total : float
        Constant kappa the closure sums to.

    Returns
    -------
    RichResult
        ``composition``, ``total``, ``D``.

    References
    ----------
    Aitchison, J. (1986), The Statistical Analysis of Compositional Data, Chapman and Hall, is this shelf's primary book and is NOT in the reference library, so it could not be read.  The definitions below were taken instead from Mateu-Figueras, G., Pawlowsky-Glahn, V. and Egozcue, J. J., The normal distribution in some constrained sample spaces, arXiv:0802.2643 (published as SORT 37(1):29-56, 2013), Sect. 4.1, which prints them with equation numbers and attributes them to Aitchison (1982, 1986); that paper was FETCHED and is archived in the reference library with a row in EXTERNAL_SOURCES.md.  The inverse of clr is the closure of the exponentiated coordinates.  It is well defined for any real z, not only for z summing to zero: adding a constant to every z_i leaves the closed result unchanged.
    """
    z = C.vec(z)
    if len(z) == 0:
        raise ValueError("z must be non-empty")
    m = max(z)
    e = [math.exp(v - m) for v in z]
    s = sum(e)
    k = float(total)
    return RichResult(payload={
        "composition": [k * v / s for v in e], "total": k, "D": len(z),
        "method": "Inverse centred log-ratio transform"})


aitchison_clr_inverse = clrinv


def cheatsheet():
    return 'aitclri: Inverse centred log-ratio transform.'

File: r-package/morie/R/aitclri.R
# SPDX-License-Identifier: AGPL-3.0-or-later
#' Map clr coordinates back to a closed composition.
#'
#' Formula: clr^-1(z) = C( exp(z_1), ..., exp(z_D) )
#'
#' @param z clr coordinates.
#' @param total Constant the returned composition sums to.
#' @return List with \code{composition}, \code{shift}, \code{total}, \code{D}.
#' @references Aitchison (1986), The Statistical Analysis of Compositional
#'   Data, Chapter 4. Verified against the reference implementation in the
#'   CRAN package compositions 2.0-9, whose clrInv is acomp(exp(z)), i.e.
#'   the closure of the exponentiated argument.
#' @export
Clrinv <- function(z, total = 1) {
  z <- .t1_vec(z)
  D <- length(z)
  shift <- sum(z) / D
  e <- exp(z - shift)
  s <- sum(e)
  k <- as.numeric(total)
  .t1_result(composition = k * e / s, shift = shift, total = k, D = D,
             method = "Inverse centred log-ratio transform")
}
